HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,87 @@
+from pyspark.accumulators import AccumulatorParam
+
+class BFSAccumulatorParam(AccumulatorParam):
+	"""
+	Defines an accumulator for the optimized search. The 
+	accumulator is a dict with key: Character, val: distances
+	"""
+	def zero(self, value):
+		return {}
+
+	def addInPlace(self, val1, val2):
+		# A way to merge two dicts and keep the min of conflicts. 
+		# Code from http://stackoverflow.com/questions/28642671/merge-dictionaries-with-minimum-value-of-common-keys
+		return {key: min(i for i in (val1.get(key), val2.get(key)) if i >= 0) for key in val1.viewkeys() | val2}
+
+def search_accum(startNode, adjacencyList, sc):
+	"""
+	The Goal of this function was to eliminate shuffle costs by 
+	using a broadcast variable. It is slightly buggy so I just went the 
+	regular way with search_bfs_new
+	"""
+	# Init the accumulator
+	accum = sc.accumulator({}, BFSAccumulatorParam())
+	# The rdd with Key: character, Val: Distances
+	distances = sc.parallelize([(startNode, 0)])
+	# A variable to keep track of nodes we have visited
+	seen = sc.broadcast(set())
+
+	# While we still have unvisited nodes
+	while not distances.isEmpty():
+		# A helper function to add ot the accumulator
+		def f(listvalue):
+			accum.add({listvalue[0]:listvalue[1]})
+
+		distances.foreach(f)
+		# Add all the elements in the rdd to to the seen set
+		current = distances.collect()
+		seen = sc.broadcast(seen.value.union(set([node[0] for node in current])))
```

Comment:
You collect and then broadcast the BFS closed list (nodes already seen) to all nodes.
This is very costly, and could be avoided by using joins, subtractByKey, and unions.